### ISSUE TITLE
Fast skipping freezes sprites animated to alpha fade in intermediate alpha, showing them half-transparent for the rest of the script

### ISSUE THREAD

**Issue by hsandt** [renpy/renpy#4790]
I use an alpha fade in when a character enters the scene:

```rpy
transform character_warp_to_middle(fade_duration=0.5):
        xpos 0.5
        ypos 1.0
        # fade in
        alpha 0.0
        easein fade_duration alpha 1.0
```

Usage ex:

```rpy
show mc neutral at character_warp_to_middle()
```

It works well! But I noticed that when I fast-skip a scene, if the characters are animated again after a few lines of dialogue, the characters will be stuck at some alpha value between 0 and 1 excluded for the rest of the script (until they are hidden again).

![screenshot0024](https://github.com/renpy/renpy/assets/1067772/d5d4c3b9-dd39-45b1-90f3-334a3662280e)

This is because Ctrl/skip will go through text so fast it will reach the next animation as the fading has not even finished (so it's not specific to Skip: if you script a fade, then a short pause (shorter than fade), then another animation moving the character, it will also interrupt the fade. The issue here is that the "bug" is not visible until you play the game once, then redo it Skipping; or enable Skip Unseen and Skip. Fortunately during debug, you Skip a lot so you're likely to spot the issue.

I thought fast skip, as well as spamming the Continue button, were automatically finishing any running animation to make sure we start next statement in the correct state.

UPDATE: I tested again and it only happens with two specific characters, after applying certain statements previously in the script. Trying a MWE.

UPDATE: managed to repro, you had to add another animation moving the character after a few lines
[Repro bug - Fast skip text to reach next interrupting animation freezes alpha.zip](https://github.com/renpy/renpy/files/11899098/Repro.bug.-.Fast.skip.text.to.reach.next.interrupting.animation.freezes.alpha.zip)

Repro:
1. Play the game once normally: OK, sprite fades in completely
2. Rollback
3. Hold Ctrl or enable Skip: note how sprite remains semi-transparent

NOTE: I made a mistake on the workaround with the commented line: it should be placed at least one statement after the alpha animation block, and at the latest before the next (moving) animation. Earlier if you really need the character to be fully visible during the next dialogue lines.

**Comment by Org-gith** [comment]
I believe that's normal and expected if you call show _thecharacter_ again before the alpha animation finishes, but that's something we can deal with...
If you want their alpha animation to finish so that they don't stuck in an intermediary value because of skipping, you could type on the next block of the alpha animation:
show character:
  alpha 1.0

But yeah if for example you had a 
show _thecharacter_ with no alpha properties following the block where you had your animation, then it would interrupt the alpha to where it was

At least that's how I deal with it, maybe could be improved or made as an option, but for now, that actually gives us more control since you can have it just like you want, either playing the animation or even if the skipping is occurring or you can just skip it on the next block by putting the alpha where you want.

**Comment by hsandt** [comment]
Indeed, I'm using this workaround for now. I even added it as commented code in the MWE.

I see how it can be useful to interpolate everything smoothly to keep the last value and interrupt previous animation: for instance, a character who has not gone completely to the left can go back to the right faster, rather than warping to the left, only to go back to the right.

But when it comes to unrelated properties, here position vs alpha, I'm not sure of the benefit. Many things can go wrong and since we move characters a lot but change brightness and alpha less often (e.g. making character not speaking darker), we are likely to end up with a bunch of characters darker and more transparent than they should be.

I'll use the workaround for now, but I need to think about all the places where I used such animations, and I cannot just integrate the fix inside the animation (there is no "finally" clause in the ATL definition), so I need to plug the fix on every usage of those transforms. Fortunately, a limited number of them change alpha and brightness.

**Comment by Org-gith** [comment]
Alternatively, to automate it partially you could do something:
transform character_warp_to_middle(fade_duration=0.5):
        xpos 0.5
        ypos 1.0
        # fade in
        alpha 0.0
        easein fade_duration*(not renpy.is_skipping()) alpha 1.0

This would make the fade_duration to be equal to 0 in duration if you are skipping.
But this would work only if the animation is not already showing (because it will be calculated once when your transform is used), so overall I prefer setting it to do the other solution, it just works for all edge cases.

Maybe I'm too conservative, but my instinct tells me that the fix would also bring some issues with it. At least here we have total control and we can do what we want at the cost of typing one additional line (once you understand it, it should not slow you down) but could be wrong.

**Comment by hsandt** [comment]
I see, it does fix the case for skipping over yes (so enough for most debugging cases), not for skipping after or if player spams Continue button to advance very fast (if there is only one statement before the next animation, this is possible).

An option to have the behavior I'm thinking of (force warp to end of animation) would help, without breaking existing projects. Otherwise, a brand new finally clause, but that's more syntax rework and would seem redundant in most cases.

I do understand we cannot always decide what the final animation state is though, if some things are dynamic, or the animation uses `repeat`. But if I use `repeat`, I'm ready to assume the sprite may have any state at the end so I'll be naturally adding some force set value safety lines anyway.

**Comment by HorBro38** [comment]
My solution is to just have another statement after the animation (in this case, a static show of the image at alpha 1.0).  This will force it to that state even if they skip.

**Comment by Gouvernathor** [comment]
Maybe you could try this
```rpy
on hide, replaced:
    alpha 1.
```

**Comment by renpytom** [comment]
This is intended behavior - we want to keep the current behavior, since it would be weird for: 

```
show eileen happy:
     alpha 0.0
     linear 4 alpha 1.0
"..."
show eileen happy:
     linear 4 alpha 0.0
"..."
```

To have alpha jump to 1.0 when the user clicks. (This would be especially weird if it was motion - you'd have things just jumping across the screen.) 

This isn't something that needs to change.

**Comment by hsandt** [comment]
> Maybe you could try this
> 
> ```python
> on hide, replaced:
>     alpha 1.
> ```

Thanks, I added this to the first transform, but the issue still occurs.

**Comment by hsandt** [comment]
> This would be especially weird if it was motion - you'd have things just jumping across the screen.

Yeah I got the point with smooth transitions. I was surprised it also happened when dealing with unrelated properties (i.e. the second transform is not changing alpha). I do understand that it may be difficult to parse all properties in advance and/or predict exact user intention. That's why I also suggested an option that developer may enable to force animation jump to end *if it makes sense in their project*. Options like this may be problematic for understanding because it makes a project behave differently on the whole, so something new debugging your code may have troubles understanding why it doesn't behave like a similar project; but something to keep in mind.

Otherwise, I found another workaround: reset properties at the start of the second transform.

In this case:

```rpy
transform character_move_to_right(duration=1.0):
        alpha 1.0  # HERE
        linear duration xpos 0.7
        ypos 1.0
```

Advantage: no fuss with local HOTFIX in script, all is done in transform
Disadvantage: only works if you are sure you want to reset property for every usage of the transform. In my case, all my characters are opaque during move. If I add a ghost later in the game, this will break, and I'll need either a variant of the transform, or, more flexible, an extra parameter alpha.

Since this happened only for alpha and is pretty easy to spot in my project by holding Skip all along, this should be enough for me. Of course, if other users stumble on more complex variants of the issue, then one of my other suggestions above may be useful.

I will rename the title to add more generic description to help users finding a similar issue.
